scapy/contrib/automotive/xcp/scanner.py

```diff
@@ -0,0 +1,148 @@
+# This file is part of Scapy
+# See http://www.secdev.org/projects/scapy for more information
+# Copyright (C) Tabea Spahn <tabea.spahn@e-mundo.de>
+# This program is published under a GPLv2 license
+
+# scapy.contrib.description = XCPScanner
+# scapy.contrib.status = loads
+
+from typing import Optional, List, Tuple
+
+from scapy.contrib.automotive.xcp.cto_commands_master import \
+    TransportLayerCmd, TransportLayerCmdGetSlaveId, Connect
+from scapy.contrib.automotive.xcp.xcp import CTORequest, XCPOnCAN
+from scapy.contrib.cansocket_native import CANSocket
+
+
+class XCPScannerResult:
+    def __init__(self, slave_id, response_id):
+        self.slave_id = slave_id
+        self.response_id = response_id
+
+
+class XCPOnCANScanner:
+    """
+    Scans for XCP Slave on CAN
+    """
+
+    def __init__(self, can_socket, id_range=None,
+                 sniff_time=0.1, verbose=False):
+        # type: (CANSocket, Optional[Tuple[int, int]], Optional[float], Optional[bool]) -> None # noqa: E501
+
+        """
+        Constructor
+        :param can_socket: Can Socket with XCPonCAN as basecls for scan
+        :param id_range: CAN id range to scan
+        :param sniff_time: time the scan waits for a response
+                           after sending a request
+        """
+        self.__socket = can_socket
+        self.id_range = id_range or (0, 0x7ff)
+        self.__flags = 0
+        self.__sniff_time = sniff_time
+        self.__verbose = verbose
+
+    def _scan(self, identifier, body, answer_type):
+        # type: (int, CTORequest, str) -> List # noqa: E501
+
+        self.log_verbose("Scan for id: " + str(identifier))
+        cto_request = \
+            XCPOnCAN(identifier=identifier, flags=self.__flags) \
+            / CTORequest() / body
+
+        req_and_res_list, _unanswered = \
+            self.__socket.sr(cto_request, timeout=self.__sniff_time,
+                             verbose=self.__verbose, multi=True)
+
+        if len(req_and_res_list) == 0:
+            self.log_verbose(
+                "No answer for identifier: " + str(identifier))
+            return []
+        valid_req_and_res_list = filter(
+            lambda req_and_res: answer_type in req_and_res[1],
```

Comment:
A comparison with instances might be better in this case

Reply:
I think instance will not work here, as it should be `CTOResponse`. I think `haslayer()` might be better here?